File: mobile/android/core/src/main/kotlin/com/portalia/core/auth/StaffSession.kt
package com.portalia.core.auth

import com.portalia.core.api.StaffApi
import com.portalia.core.config.AppConfig
import com.portalia.core.model.StaffMeResponse
import com.portalia.core.net.ApiError
import com.portalia.core.net.HttpClient
import kotlinx.coroutines.Deferred
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import okhttp3.HttpUrl.Companion.toHttpUrl

/**
 * Sessão do app do porteiro (equipe). Espelha StaffSession do iOS: access
 * token só em memória, refresh cookie httpOnly persiste via
 * PersistentCookieJar (mesmo papel que HTTPCookieStorage no iOS).
 */
class StaffSession(
    val api: StaffApi,
    private val cookieJar: PersistentCookieJar,
) {
    private val _me = MutableStateFlow<StaffMeResponse?>(null)
    val me: StateFlow<StaffMeResponse?> = _me

    private val _mustChangePassword = MutableStateFlow(false)
    val mustChangePassword: StateFlow<Boolean> = _mustChangePassword

    private val _isBootstrapping = MutableStateFlow(true)
    val isBootstrapping: StateFlow<Boolean> = _isBootstrapping

    private val _errorMessage = MutableStateFlow<String?>(null)
    val errorMessage: StateFlow<String?> = _errorMessage

    private var accessToken: String? = null
    private var csrfToken: String? = null
    private var inFlightRefresh: Deferred<Unit>? = null

    val isAuthenticated: Boolean get() = accessToken != null && _me.value != null
    val condominiumId: String? get() = _me.value?.condominiumId

    suspend fun bootstrap() {
        refresh()
        _isBootstrapping.value = false
    }

    suspend fun login(email: String, password: String): Boolean {
        _errorMessage.value = null
        return try {
            val result = api.login(email, password)
            accessToken = result.accessToken
            csrfToken = result.csrfToken
            _mustChangePassword.value = result.mustChangePassword
            loadMe()
            true
        } catch (e: Exception) {
            _errorMessage.value = (e as? ApiError)?.message ?: "Não foi possível entrar."
            false
        }
    }

    suspend fun logout() {
        val token = accessToken
        val csrf = csrfToken
        if (token != null && csrf != null) {
            try { api.logout(token, csrf) } catch (e: Exception) { /* já saindo mesmo assim */ }
        }
        cookieJar.clear()
        accessToken = null
        csrfToken = null
        _me.value = null
        _mustChangePassword.value = false
    }

    suspend fun changePassword(current: String, new: String): Boolean {
        val token = accessToken ?: return false
        _errorMessage.value = null
        return try {
            val result = api.changePassword(token, current, new)
            accessToken = result.accessToken
            csrfToken = result.csrfToken
            _mustChangePassword.value = false
            true
        } catch (e: Exception) {
            _errorMessage.value = (e as? ApiError)?.message ?: "Não foi possível trocar a senha."
            false
        }
    }

    suspend fun <T> authorized(call: suspend (String) -> T): T {
        val token = accessToken ?: throw ApiError.Unauthorized
        return try {
            call(token)
        } catch (e: ApiError.Unauthorized) {
            refresh()
            val retryToken = accessToken ?: throw ApiError.Unauthorized
            call(retryToken)
        }
    }

    /**
     * Chamadores concorrentes (ex.: duas telas relançando `authorized{}` ao
     * mesmo tempo depois de um 401) compartilham a mesma corrida de
     * refresh em vez de cada uma disparar a sua — mesmo desenho do
     * `inFlightRefresh` do frontend web e do iOS. Assume que este objeto é
     * sempre acessado a partir do dispatcher principal (Compose/ViewModel),
     * então não há necessidade de um Mutex explícito aqui.
     */
    private suspend fun refresh() {
        val existing = inFlightRefresh
        if (existing != null) {
            existing.await()
            return
        }
        coroutineScope {
            val task = async { performRefresh() }
            inFlightRefresh = task
            task.await()
            inFlightRefresh = null
        }
    }

    private suspend fun performRefresh() {
        val url = AppConfig.API_BASE_URL.toHttpUrl()
        val tokenToUse = csrfToken ?: cookieJar.cookieValue(url, "portalia_csrf")
        try {
            val result = api.refresh(tokenToUse ?: "")
            accessToken = result.accessToken
            csrfToken = result.csrfToken
            _mustChangePassword.value = result.mustChangePassword
            if (_me.value == null) loadMe()
        } catch (e: Exception) {
            accessToken = null
            csrfToken = null
            _me.value = null
        }
    }

    private suspend fun loadMe() {
        val token = accessToken ?: return
        try {
            _me.value = api.me(token)
        } catch (e: Exception) {
            _errorMessage.value = (e as? ApiError)?.message
        }
    }

    companion object {
        fun create(context: android.content.Context): StaffSession {
            val jar = PersistentCookieJar(SecureStore.shared(context))
            val http = HttpClient(cookieJar = jar)
            return StaffSession(StaffApi(http), jar)
        }
    }
}

File: mobile/ios/PortaliaCore/Sources/PortaliaCore/Auth/StaffSession.swift
import Foundation

/// Sessão do app do porteiro (equipe). Espelha o session-store do frontend
/// web: access token só em memória (Published, nunca persistido em claro —
/// só o refresh cookie httpOnly persiste, gerenciado pelo próprio
/// URLSession), com uma tentativa silenciosa de refresh ao abrir o app.
@MainActor
public final class StaffSession: ObservableObject {
    @Published public private(set) var me: StaffMeResponse?
    @Published public private(set) var mustChangePassword = false
    @Published public private(set) var isBootstrapping = true
    @Published public var errorMessage: String?

    public let api: StaffAPI
    private var accessToken: String?
    private var csrfToken: String?
    private var inFlightRefresh: Task<Void, Never>?

    public var isAuthenticated: Bool { accessToken != nil && me != nil }
    public var condominiumId: String? { me?.condominiumId }

    public init(api: StaffAPI = StaffAPI(http: HTTPClient())) {
        self.api = api
    }

    public func bootstrap() async {
        await refresh()
        isBootstrapping = false
    }

    public func login(email: String, password: String) async -> Bool {
        errorMessage = nil
        do {
            let result = try await api.login(email: email, password: password)
            accessToken = result.accessToken
            csrfToken = result.csrfToken
            mustChangePassword = result.mustChangePassword
            await loadMe()
            return true
        } catch {
            errorMessage = (error as? APIError)?.errorDescription ?? "Não foi possível entrar."
            return false
        }
    }

    public func logout() async {
        if let accessToken, let csrfToken {
            try? await api.logout(accessToken: accessToken, csrfToken: csrfToken)
        }
        self.accessToken = nil
        self.csrfToken = nil
        self.me = nil
        self.mustChangePassword = false
    }

    public func changePassword(current: String, new: String) async -> Bool {
        guard let accessToken else { return false }
        errorMessage = nil
        do {
            let result = try await api.changePassword(accessToken: accessToken, current: current, new: new)
            self.accessToken = result.accessToken
            self.csrfToken = result.csrfToken
            self.mustChangePassword = false
            return true
        } catch {
            errorMessage = (error as? APIError)?.errorDescription ?? "Não foi possível trocar a senha."
            return false
        }
    }

    /// Executa uma chamada autenticada; numa resposta 401, tenta um refresh
    /// (compartilhado entre chamadores concorrentes, mesmo desenho do
    /// `inFlightRefresh` do frontend web) e tenta de novo uma única vez.
    public func authorized<T>(_ call: (String) async throws -> T) async throws -> T {
        guard let token = accessToken else { throw APIError.unauthorized }
        do {
            return try await call(token)
        } catch APIError.unauthorized {
            await refresh()
            guard let retryToken = accessToken else { throw APIError.unauthorized }
            return try await call(retryToken)
        }
    }

    private func refresh() async {
        if let inFlightRefresh {
            await inFlightRefresh.value
            return
        }
        // Tipo anotado explicitamente: `self?.performRefresh()` devolve
        // `Void?` (opcional encadeado), então sem a anotação o Swift infere
        // `Task<Void?, Never>`, que não casa com `Task<Void, Never>?`.
        let task: Task<Void, Never> = Task { [weak self] in
            await self?.performRefresh()
        }
        inFlightRefresh = task
        await task.value
        inFlightRefresh = nil
    }

    /// Se o app acabou de abrir (relançado depois de fechado), ainda não há
    /// `csrfToken` em memória — mas o cookie CSRF (não-httpOnly, de propósito)
    /// pode ter sobrevivido no `HTTPCookieStorage` persistente do URLSession
    /// de uma sessão anterior. Lê de lá como fallback só nesse caso.
    private func performRefresh() async {
        let tokenToUse = csrfToken ?? readCsrfCookieFromStorage()
        do {
            let result = try await api.refresh(csrfToken: tokenToUse ?? "")
            self.accessToken = result.accessToken
            self.csrfToken = result.csrfToken
            self.mustChangePassword = result.mustChangePassword
            if me == nil { await loadMe() }
        } catch {
            self.accessToken = nil
            self.csrfToken = nil
            self.me = nil
        }
    }

    private func readCsrfCookieFromStorage() -> String? {
        // O backend grava o cookie com `Path=/api/v1/auth`, e o casamento de
        // path de cookie exige que o path do cookie seja prefixo do path
        // consultado — procurar na base `/api/v1` nunca encontraria nada.
        let authURL = AppConfig.apiBaseURL.appendingPathComponent("auth")
        return HTTPCookieStorage.shared.cookies(for: authURL)?
            .first(where: { $0.name == "portalia_csrf" })?.value
    }

    private func loadMe() async {
        guard let accessToken else { return }
        do {
            me = try await api.me(accessToken: accessToken)
        } catch {
            errorMessage = (error as? APIError)?.errorDescription
        }
    }
}
